tf/tfprocess.py

```diff
@@ -135,10 +135,15 @@ def init_net(self, next_batch):
         self.batch_norm_count = 0
         self.y_conv, self.z_conv = self.construct_net(self.x)
 
-        # y_ has -1 on illegal moves, flush them to 0 first
-        self.y_ = tf.nn.relu(self.y_)
-
         # Calculate loss on policy head
+        if self.cfg['training'].get('mask_legal_moves'):
+            # extract mask for legal moves from target policy
+            move_is_legal = tf.greater_equal(self.y_, 0)
+            # replace logits of illegal moves with large negative value (so that it doesn't affect policy of legal moves) without gradient
+            illegal_filler = tf.zeros_like(self.y_conv) - 1.0e-10
```

Comment:
1e-10 is not very large value. I guess it should be 1e10?

Any possible numerical issues from using that large value? If Tensorflow calculates sum of policy, using large value for illegal moves can cause loss of precision.

Reply:
Whoops, indeed.

There shouldn't be any numerical issues. When the large negative value goes into the denominator of the softmax, it becomes virtually 0 because it appears as an exponent, so it doesn't affect the sum over exponentials in the denominator. When it appears in the nominator, it's multiplied by the 0 probability in the target vector when computing the loss, before the loss is reduced to a scalar, so again it doesn't affect anything.